File: package/ios/CameraError.swift
//
//  CameraError.swift
//  mrousavy
//
//  Created by Marc Rousavy on 14.01.21.
//  Copyright © 2021 mrousavy. All rights reserved.
//

import Foundation

// MARK: - PermissionError

enum PermissionError: String {
  case microphone = "microphone-permission-denied"
  case camera = "camera-permission-denied"

  var code: String {
    return rawValue
  }

  var message: String {
    switch self {
    case .microphone:
      return "The Microphone permission was denied! If you want to record Videos without sound, pass `audio={false}`."
    case .camera:
      return "The Camera permission was denied!"
    }
  }
}

// MARK: - ParameterError

enum ParameterError {
  case invalid(unionName: String, receivedValue: String)
  case unsupportedOS(unionName: String, receivedValue: String, supportedOnOs: String)
  case unsupportedOutput(outputDescriptor: String)
  case unsupportedInput(inputDescriptor: String)
  case invalidCombination(provided: String, missing: String)

  var code: String {
    switch self {
    case .invalid:
      return "invalid-parameter"
    case .unsupportedOS:
      return "unsupported-os"
    case .unsupportedOutput:
      return "unsupported-output"
    case .unsupportedInput:
      return "unsupported-input"
    case .invalidCombination:
      return "invalid-combination"
    }
  }

  var message: String {
    switch self {
    case let .invalid(unionName: unionName, receivedValue: receivedValue):
      return "The value \"\(receivedValue)\" could not be parsed to type \(unionName)!"
    case let .unsupportedOS(unionName: unionName, receivedValue: receivedValue, supportedOnOs: os):
      return "The value \"\(receivedValue)\" for type \(unionName) is not supported on the current iOS version! Required OS: \(os) or higher"
    case let .unsupportedOutput(outputDescriptor: output):
      return "The output \"\(output)\" is not supported!"
    case let .unsupportedInput(inputDescriptor: input):
      return "The input \"\(input)\" is not supported!"
    case let .invalidCombination(provided: provided, missing: missing):
      return "Invalid combination! If \"\(provided)\" is provided, \"\(missing)\" also has to be set!"
    }
  }
}

// MARK: - DeviceError

enum DeviceError: String {
  case configureError = "configuration-error"
  case noDevice = "no-device"
  case invalid = "invalid-device"
  case flashUnavailable = "flash-unavailable"
  case microphoneUnavailable = "microphone-unavailable"
  case lowLightBoostNotSupported = "low-light-boost-not-supported"
  case focusNotSupported = "focus-not-supported"
  case notAvailableOnSimulator = "camera-not-available-on-simulator"
  case pixelFormatNotSupported = "pixel-format-not-supported"

  var code: String {
    return rawValue
  }

  var message: String {
    switch self {
    case .configureError:
      return "Failed to lock the device for configuration."
    case .noDevice:
      return "No device was set! Use `useCameraDevice(..)` or `Camera.getAvailableCameraDevices()` to select a suitable Camera device."
    case .invalid:
      return "The given Camera device was invalid. Use `useCameraDevice(..)` or `Camera.getAvailableCameraDevices()` to select a suitable Camera device."
    case .flashUnavailable:
      return "The Camera Device does not have a flash unit! Make sure you select a device where `hasFlash`/`hasTorch` is true!"
    case .lowLightBoostNotSupported:
      return "The currently selected camera device does not support low-light boost! Make sure you select a device where `supportsLowLightBoost` is true!"
    case .focusNotSupported:
      return "The currently selected camera device does not support focussing!"
    case .microphoneUnavailable:
      return "The microphone was unavailable."
    case .notAvailableOnSimulator:
      return "The Camera is not available on the iOS Simulator!"
    case .pixelFormatNotSupported:
      return "The given pixelFormat is not supported on the given Camera Device!"
    }
  }
}

// MARK: - FormatError

enum FormatError {
  case invalidFps(fps: Int)
  case invalidHdr
  case invalidFormat
  case incompatiblePixelFormatWithHDR

  var code: String {
    switch self {
    case .invalidFormat:
      return "invalid-format"
    case .invalidFps:
      return "invalid-fps"
    case .invalidHdr:
      return "invalid-hdr"
    case .incompatiblePixelFormatWithHDR:
      return "incompatible-pixel-format-with-hdr-setting"
    }
  }

  var message: String {
    switch self {
    case .invalidFormat:
      return "The given format was invalid. Did you check if the current device supports the given format in `device.formats`?"
    case let .invalidFps(fps):
      return "The given format cannot run at \(fps) FPS! Make sure your FPS is lower than `format.maxFps` but higher than `format.minFps`."
    case .invalidHdr:
      return "The currently selected format does not support HDR capture! Make sure you select a format which includes `supportsPhotoHDR`/`supportsVideoHDR`!"
    case .incompatiblePixelFormatWithHDR:
      return "The currently selected pixelFormat is not compatible with HDR! HDR only works with the `yuv` pixelFormat."
    }
  }
}

// MARK: - SessionError

enum SessionError {
  case cameraNotReady
  case audioSessionSetupFailed(reason: String)
  case audioSessionFailedToActivate
  case audioInUseByOtherApp

  var code: String {
    switch self {
    case .cameraNotReady:
      return "camera-not-ready"
    case .audioSessionSetupFailed:
      return "audio-session-setup-failed"
    case .audioInUseByOtherApp:
      return "audio-in-use-by-other-app"
    case .audioSessionFailedToActivate:
      return "audio-session-failed-to-activate"
    }
  }

  var message: String {
    switch self {
    case .cameraNotReady:
      return "The Camera is not ready yet! Wait for the onInitialized() callback!"
    case let .audioSessionSetupFailed(reason):
      return "The audio session failed to setup! \(reason)"
    case .audioInUseByOtherApp:
      return "The audio session is already in use by another app with higher priority!"
    case .audioSessionFailedToActivate:
      return "Failed to activate Audio Session!"
    }
  }
}

// MARK: - CaptureError

enum CaptureError {
  case invalidPhotoFormat
  case recordingInProgress
  case noRecordingInProgress
  case fileError
  case createTempFileError
  case createRecorderError(message: String? = nil)
  case invalidPhotoCodec
  case videoNotEnabled
  case photoNotEnabled
  case aborted
  case unknown(message: String? = nil)

  var code: String {
    switch self {
    case .invalidPhotoFormat:
      return "invalid-photo-format"
    case .recordingInProgress:
      return "recording-in-progress"
    case .noRecordingInProgress:
      return "no-recording-in-progress"
    case .fileError:
      return "file-io-error"
    case .createTempFileError:
      return "create-temp-file-error"
    case .createRecorderError:
      return "create-recorder-error"
    case .invalidPhotoCodec:
      return "invalid-photo-codec"
    case .videoNotEnabled:
      return "video-not-enabled"
    case .photoNotEnabled:
      return "photo-not-enabled"
    case .aborted:
      return "aborted"
    case .unknown:
      return "unknown"
    }
  }

  var message: String {
    switch self {
    case .invalidPhotoFormat:
      return "The given photo format was invalid!"
    case .invalidPhotoCodec:
      return "The given photo codec was invalid!"
    case .recordingInProgress:
      return "There is already an active video recording in progress! Did you call startRecording() twice?"
    case .noRecordingInProgress:
      return "There was no active video recording in progress! Did you call stopRecording() twice?"
    case .fileError:
      return "An unexpected File IO error occured!"
    case .createTempFileError:
      return "Failed to create a temporary file!"
    case let .createRecorderError(message: message):
      return "Failed to create the AVAssetWriter (Recorder)! \(message ?? "(no additional message)")"
    case .videoNotEnabled:
      return "Video capture is disabled! Pass `video={true}` to enable video recordings."
    case .photoNotEnabled:
      return "Photo capture is disabled! Pass `photo={true}` to enable photo capture."
    case .aborted:
      return "The capture has been stopped before any input data arrived."
    case let .unknown(message: message):
      return message ?? "An unknown error occured while capturing a video/photo."
    }
  }
}

// MARK: - SystemError

enum SystemError: String {
  case noManager = "no-camera-manager"
  case frameProcessorsUnavailable = "frame-processors-unavailable"

  var code: String {
    return rawValue
  }

  var message: String {
    switch self {
    case .noManager:
      return "No Camera Manager was found."
    case .frameProcessorsUnavailable:
      return "Frame Processors are unavailable - is react-native-worklets-core installed?"
    }
  }
}

// MARK: - CameraError

enum CameraError: Error {
  case permission(_ id: PermissionError)
  case parameter(_ id: ParameterError)
  case device(_ id: DeviceError)
  case format(_ id: FormatError)
  case session(_ id: SessionError)
  case capture(_ id: CaptureError)
  case system(_ id: SystemError)
  case unknown(message: String? = nil)

  var code: String {
    switch self {
    case let .permission(id: id):
      return "permission/\(id.code)"
    case let .parameter(id: id):
      return "parameter/\(id.code)"
    case let .device(id: id):
      return "device/\(id.code)"
    case let .format(id: id):
      return "format/\(id.code)"
    case let .session(id: id):
      return "session/\(id.code)"
    case let .capture(id: id):
      return "capture/\(id.code)"
    case let .system(id: id):
      return "system/\(id.code)"
    case .unknown:
      return "unknown/unknown"
    }
  }

  var message: String {
    switch self {
    case let .permission(id: id):
      return id.message
    case let .parameter(id: id):
      return id.message
    case let .device(id: id):
      return id.message
    case let .format(id: id):
      return id.message
    case let .session(id: id):
      return id.message
    case let .capture(id: id):
      return id.message
    case let .system(id: id):
      return id.message
    case let .unknown(message: message):
      return message ?? "An unexpected error occured."
    }
  }
}

File: android/src/main/java/com/mrousavy/camera/core/CameraError.kt
package com.mrousavy.camera.core

import com.mrousavy.camera.core.types.VideoStabilizationMode

abstract class CameraError(
  /**
   * The domain of the error. Error domains are used to group errors.
   *
   * Example: "permission"
   */
  val domain: String,
  /**
   * The id of the error. Errors are uniquely identified under a given domain.
   *
   * Example: "microphone-permission-denied"
   */
  val id: String,
  /**
   * A detailed error description of "what went wrong".
   *
   * Example: "The microphone permission was denied!"
   */
  override val message: String,
  /**
   * A throwable that caused this error.
   */
  cause: Throwable? = null
) : Throwable("[$domain/$id] $message", cause) {
  val code: String
    get() = "$domain/$id"
}

class CameraPermissionError : CameraError("permission", "camera-permission-denied", "The Camera permission was denied!")
class MicrophonePermissionError :
  CameraError(
    "permission",
    "microphone-permission-denied",
    "The Microphone permission was denied! If you want to record Video without sound, pass `audio={false}`."
  )
class LocationPermissionError :
  CameraError(
    "permission",
    "location-permission-denied",
    "The Location permission was denied! If you want to capture photos or videos without location tags, pass `enableLocation={false}`."
  )

class InvalidTypeScriptUnionError(unionName: String, unionValue: String?) :
  CameraError("parameter", "invalid-parameter", "The given value for $unionName could not be parsed! (Received: $unionValue)")

class NoCameraDeviceError :
  CameraError(
    "device",
    "no-device",
    "No device was set! Use `useCameraDevice(..)` or `Camera.getAvailableCameraDevices()` to select a suitable Camera device."
  )
class PixelFormatNotSupportedError(format: String) :
  CameraError("device", "pixel-format-not-supported", "The pixelFormat $format is not supported on the given Camera Device!")
class FlashUnavailableError :
  CameraError(
    "device",
    "flash-not-available",
    "The Camera Device does not have a flash unit! Make sure you select a device where `device.hasFlash`/`device.hasTorch` is true."
  )
class FocusNotSupportedError :
  CameraError("device", "focus-not-supported", "The currently selected camera device does not support focusing!")
class CameraInUseError(cause: Throwable?) :
  CameraError("device", "camera-already-in-use", "The given Camera Device is already in use!", cause)
class FatalCameraError(cause: Throwable?) :
  CameraError("device", "fatal-error", "An unknown fatal error occurred in the Camera HAL! Try restarting the phone.", cause)

class CameraNotReadyError :
  CameraError("session", "camera-not-ready", "The Camera is not ready yet! Wait for the onInitialized() callback!")
class NoOutputsError :
  CameraError("session", "no-outputs", "Cannot create a CameraCaptureSession without any outputs! (PREVIEW, PHOTO, VIDEO, ...)")
class RecoverableError(cause: Throwable?) :
  CameraError(
    "session",
    "recoverable-error",
    "An unknown error occurred while creating the Camera Session, but the Camera can recover from it.",
    cause
  )
class InvalidOutputConfigurationError(cause: Throwable?) :
  CameraError(
    "session",
    "invalid-output-configuration",
    "Failed to configure the Camera Session because the output/stream configurations are invalid!",
    cause
  )

class PropRequiresFormatToBeNonNullError(propName: String) :
  CameraError("format", "format-required", "The prop \"$propName\" requires a format to be set, but format was null!")
class InvalidFpsError(fps: Int) :
  CameraError(
    "format",
    "invalid-fps",
    "The given format cannot run at $fps FPS! Make sure your FPS is lower than `format.maxFps` but higher than `format.minFps`."
  )
class InvalidVideoStabilizationMode(mode: VideoStabilizationMode) :
  CameraError(
    "format",
    "invalid-video-stabilization-mode",
    "The given format does not support the videoStabilizationMode \"${mode.unionValue}\"! " +
      "Select a format that contains ${mode.unionValue} in `format.supportedVideoStabilizationModes`."
  )
class InvalidVideoHdrError :
  CameraError(
    "format",
    "invalid-video-hdr",
    "The given format does not support videoHdr! Select a format where `format.supportsVideoHdr` is true."
  )
class PhotoHdrAndVideoHdrNotSupportedSimultaneously :
  CameraError(
    "format",
    "photo-hdr-and-video-hdr-not-suppoted-simultaneously",
    "Photo HDR and Video HDR are not supported simultaneously! Disable either `videoHdr` or `photoHdr`."
  )
class LowLightBoostNotSupportedWithHdr :
  CameraError(
    "format",
    "low-light-boost-not-supported-with-hdr",
    "The low light boost extension does not work when HDR is enabled! Disable either `lowLightBoost` or `videoHdr`/`photoHdr`."
  )

class VideoNotEnabledError :
  CameraError("capture", "video-not-enabled", "Video capture is disabled! Pass `video={true}` to enable video recordings.")
class PhotoNotEnabledError :
  CameraError("capture", "photo-not-enabled", "Photo capture is disabled! Pass `photo={true}` to enable photo capture.")
class SnapshotFailedError :
  CameraError("capture", "snapshot-failed", "Failed to take a Snapshot of the Preview View! Try using takePhoto() instead.")
class SnapshotFailedPreviewNotEnabledError :
  CameraError(
    "capture",
    "snapshot-failed-preview-not-enabled",
    "Failed to take a Snapshot because preview={...} was disabled! " +
      "Enable preview and try again."
  )
class FocusCanceledError : CameraError("capture", "focus-canceled", "The focus operation has been canceled by a new focus request.")
class FocusRequiresPreviewError : CameraError("capture", "focus-requires-preview", "Focus requires preview={...} to be enabled!")

private fun getVideoCapturedMessage(wasVideoCaptured: Boolean): String =
  if (wasVideoCaptured) {
    "The output file was generated, so the recording may be valid."
  } else {
    "The output file was generated but the recording will not be valid, so you should delete the file."
  }

open class RecorderError(id: String, message: String, val wasVideoRecorded: Boolean, cause: Throwable?) :
  CameraError("capture", id, message, cause)
class UnknownRecorderError(wasVideoRecorded: Boolean, cause: Throwable?) :
  RecorderError(
    "recorder-error",
    "An error occurred while recording a video! ${getVideoCapturedMessage(wasVideoRecorded)} ${cause?.message}",
    wasVideoRecorded,
    cause
  )
class EncoderError(cause: Throwable?) :
  RecorderError("encoder-error", "The Video Encoder encountered an error occurred while recording a video!", false, cause)
class NoDataError(cause: Throwable?) :
  RecorderError(
    "no-data",
    "The Video Recording failed because no data was received! (${cause?.message}) " +
      "Did you stop the recording before any Frames arrived?",
    false,
    cause
  )
class InvalidRecorderConfigurationError(cause: Throwable?) :
  RecorderError(
    "invalid-recorder-configuration",
    "The Video Recording failed because it was configured with invalid settings! ${cause?.message}",
    false,
    cause
  )
class FileSizeLimitReachedError(cause: Throwable?) :
  RecorderError(
    "file-size-limit-reached",
    "The Video Recording was stopped because the file size limit was reached. The output file may still be valid.",
    true,
    cause
  )
class DurationLimitReachedError(cause: Throwable?) :
  RecorderError(
    "duration-limit-reached",
    "The Video Recording was stopped because the duration limit was reached. The output file may still be valid.",
    true,
    cause
  )
class InsufficientStorageForRecorderError(cause: Throwable?) :
  RecorderError("insufficient-storage", "There is not enough storage space available for a Video Recording.", false, cause)
class NoRecordingInProgressError :
  CameraError("capture", "no-recording-in-progress", "There was no active video recording in progress! Did you call stopRecording() twice?")
class RecordingCanceledError : CameraError("capture", "recording-canceled", "The active recording was canceled.")
class FileIOError(throwable: Throwable) :
  CameraError("capture", "file-io-error", "An unexpected File IO error occurred! Error: ${throwable.message}.", throwable)
class InvalidPathError(path: String) : CameraError("capture", "invalid-path", "The given path ($path) is invalid, or not writable!")
class RecordingInProgressError :
  CameraError(
    "capture",
    "recording-in-progress",
    "There is already an active video recording in progress! Did you call startRecording() twice?"
  )
class FrameInvalidError :
  CameraError(
    "capture",
    "frame-invalid",
    "Trying to access an already closed Frame! " +
      "Are you trying to access the Image data outside of a Frame Processor's lifetime?\n" +
      "- If you want to use `console.log(frame)`, use `console.log(frame.toString())` instead.\n" +
      "- If you want to do async processing, use `runAsync(...)` instead.\n" +
      "- If you want to use runOnJS, increment it's ref-count: `frame.incrementRefCount()`"
  )
class InvalidImageTypeError : CameraError("capture", "invalid-image-type", "Captured an Image with an invalid Image type!")

class CodeTypeNotSupportedError(codeType: String) :
  CameraError(
    "code-scanner",
    "code-type-not-supported",
    "The codeType \"$codeType\" is not supported by the Code Scanner!"
  )

class ViewNotFoundError(viewId: Int) :
  CameraError("system", "view-not-found", "The given view (ID $viewId) was not found in the view manager.")
class HardwareBuffersNotAvailableError :
  CameraError("system", "hardware-buffers-unavailable", "HardwareBuffers are only available on API 28 or higher!")
class MaxCamerasInUseError(cause: Throwable?) :
  CameraError("system", "max-cameras-in-use", "The maximum amount of Cameras available for simultaneous use has been reached!", cause)
class CameraIsRestrictedError(cause: Throwable?) :
  CameraError(
    "system",
    "camera-is-restricted",
    "Camera functionality is not available because it has been restricted by the operating system, possibly due to a device policy.",
    cause
  )
class DoNotDisturbBugError(cause: Throwable?) :
  CameraError(
    "system",
    "do-not-disturb-bug",
    "The Camera Device could not be opened because of a bug in Android 9 (API 28) when do-not-disturb mode is enabled! " +
      "Either update your Android version, or disable do-not-disturb.",
    cause
  )
class RecordingWhileFrameProcessingUnavailable :
  CameraError(
    "system",
    "recording-while-frame-processing-unavailable",
    "Video Recordings are not possible with a Frame Processor running, " +
      "because the device is running on API 22 or lower and ImageWriters are not available."
  )

class UnknownCameraError(cause: Throwable?) : CameraError("unknown", "unknown", cause?.message ?: "An unknown camera error occured.", cause)
